low-level/sensors/impl/centos_7/service_monitor.py

```diff
@@ -231,19 +246,24 @@ def properties_changed_handler(self, interface, changed_properties,
         if self.state != self.previous_state:
             if self.state == "active":
                 if self._service_state is FailedState:
+                    self.waiting_timestamp = time.time()
+                    Service.active_waiting_services.add(self.name)
+                elif self._service_state is InactiveState:
                     self.new_service_state(ActiveState)
-                    Service.active_services.add(self.name)
-                if self._service_state is InactiveState:
-                    self.new_service_state(ActiveState)
+                # TODO: Handle discarding service from state list
+                # in new_service_state function.
+                Service.failed_waiting_services.discard(self.name)
             elif self.state == "failed":
                 if self._service_state is not FailedState:
-                    Service.alerts.put(
-                        ServiceMonitor.get_alert(self, FailedAlert))
-                self.new_service_state(FailedState)
+                    self.waiting_timestamp = time.time()
+                    Service.failed_waiting_services.add(self.name)
+                Service.active_waiting_services.discard(self.name)
             else:
                 # Avoid raising duplicate alerts
                 if self._service_state != FailedState:
                     self.new_service_state(InactiveState)
+                Service.active_waiting_services.discard(self.name)
+                Service.failed_waiting_services.discard(self.name)
```

Comment:
similar way if we add `InactiveWaiting` state then actions under if-else will reduce,
*_But after adding these states we will again have to calculate transitions and verify actions so **I am good with the current changes**._